packages/remix-react/__tests__/transition-test.tsx

```diff
@@ -2687,6 +2687,38 @@ describe("deferred", () => {
     // cannot be opted out of
     expect(shouldReloadSpy).toHaveBeenCalledTimes(1);
   });
+
+  it("removes pending deferreds on successful resolution to re-enable shouldReload on fetch submissions", async () => {
+    let shouldReloadSpy = jest.fn(() => false);
+    let t = setup({ url: "/", shouldReloads: { foo: shouldReloadSpy } });
+
+    let A = t.navigate.get("/foo");
+    let dfdA = defer();
+    await A.loader.resolve(
+      setupDeferred({
+        value: dfdA.promise,
+      })
+    );
+    await dfdA.resolve("A");
+    expect(t.getState().loaderData).toEqual({
+      root: "ROOT",
+      foo: {
+        value: "A",
+      },
+    });
+    expect(shouldReloadSpy).toHaveBeenCalledTimes(0);
+
+    // Navigate to the same URL forcing a revalidation, which foo can opt out of
+    let B = t.fetch.post("/bar");
+    await B.action.resolve("FETCH ACTION");
+    expect(t.getState().loaderData).toEqual({
+      root: "ROOT",
+      foo: {
+        value: "A",
+      },
+    });
+    expect(shouldReloadSpy).toHaveBeenCalledTimes(1);
```

Comment:
This is the main aspect here - `shouldReload` was never getting called on a fetcher submission _after_ a resolved deferred